extensions/aris/src/bigDataCluster/tree/treeNode.ts

```diff
@@ -0,0 +1,248 @@
+		return TreeNode.findNode(this, condition, expandIfNeeded);
+	}
+
+	public static async findNode(node: TreeNode, condition: (node: TreeNode) => boolean, expandIfNeeded?: boolean): Promise<TreeNode> {
+		if (!node || !condition) {
+			return undefined;
+		}
+		let result: TreeNode = undefined;
+		let nodesToCheck: TreeNode[] = [node];
+		while (nodesToCheck.length > 0) {
+			let current = nodesToCheck.shift();
+			if (condition(current)) {
+				result = current;
+				break;
+			}
+			if (current.hasChildren) {
+				nodesToCheck = nodesToCheck.concat(current.children);
+			} else if (!current.isLeaf && expandIfNeeded) {
+				nodesToCheck = nodesToCheck.concat(await current.getChildren());
+			}
+		}
+		return result;
+	}
+
+	public async filterNode(condition: (node: TreeNode) => boolean, expandIfNeeded?: boolean): Promise<TreeNode[]> {
+		return TreeNode.filterNode(this, condition, expandIfNeeded);
+	}
+
+	public static async filterNode(node: TreeNode, condition: (node: TreeNode) => boolean, expandIfNeeded?: boolean): Promise<TreeNode[]> {
+		if (!node || !condition) {
+			return undefined;
+		}
+		let result: TreeNode[] = [];
+		let nodesToCheck: TreeNode[] = [node];
+		while (nodesToCheck.length > 0) {
+			let current = nodesToCheck.shift();
+			if (condition(current)) {
+				result.push(current);
+			}
+			if (current.hasChildren) {
+				nodesToCheck = nodesToCheck.concat(current.children);
+			} else if (!current.isLeaf && expandIfNeeded) {
+				nodesToCheck = nodesToCheck.concat(await current.getChildren());
+			}
+		}
+		return result;
+	}
+
+	public async findNodeByPath(path: string, expandIfNeeded?: boolean): Promise<TreeNode> {
+		return TreeNode.findNodeByPath(this, path, expandIfNeeded);
+	}
+
+	public static async findNodeByPath(node: TreeNode, path: string, expandIfNeeded?: boolean): Promise<TreeNode> {
+		return TreeNode.findNode(node, node => {
+			return node.nodePath && (node.nodePath === path || node.nodePath.startsWith(path));
+		}, expandIfNeeded);
+	}
+
+	public async getChildren(refresh?: boolean): Promise<TreeNode[]> {
+		if ((!this.hasChildren && !this._isLeaf) || refresh) {
+			await this.expand().then(children => {
+				this._children = children;
+				if (this._autoLeaf) {
+					this._isLeaf = !this.hasChildren;
+				}
+			}, error => {
+				this._isLeaf = false;
+			});
+		}
+		return this._children;
+	}
+
+	public async expand(): Promise<TreeNode[]> {
+		return [];
+	}
+
+	public addChild(node: TreeNode): void {
+		if (!this._children) {
+			this._children = [];
+		}
+		this._children.push(node);
+		if (this._autoLeaf) {
+			this._isLeaf = false;
+		}
+	}
+
+	public clearChildren(): void {
+		if (this._children) {
+			this._children = [];
+			if (this._autoLeaf) {
+				this._isLeaf = false;
+			}
+		}
+	}
+
+	public abstract getTreeItem(): vscode.TreeItem;
+	public abstract getNodeInfo(): azdata.NodeInfo;
+
+	private static generateGuid(): string {
```

Comment:
>generateGuid [](start = 16, length = 12)

This seems like something that should go in a general utils class

Reply:
moved.